plugins/nitrokey/fu-plugin-nitrokey.c

```diff
@@ -0,0 +1,365 @@
+
+	g_return_val_if_fail (buf_in_sz <= 59, FALSE);
+	g_return_val_if_fail (buf_out_sz <= 54, FALSE);
+
+	memset (&req, 0x00, sizeof(req));
+	req.command = command;
+	crc_tmp = _stm_crc32 ((const guint8 *) &req, sizeof(req) - 4);
+	req.crc = GUINT32_TO_LE (crc_tmp);
+	if (buf_in != NULL)
+		memcpy (&req.payload, buf_in, buf_in_sz);
+
+	/* send request */
+	_dump_to_console ("request", (const guint8 *) &req, sizeof(req));
+	ret = g_usb_device_control_transfer (usb_device,
+					     G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					     G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					     G_USB_DEVICE_RECIPIENT_INTERFACE,
+					     0x09,
+					     0x0300,
+					     0x0002,
+					     (const guint8 *) &req,
+					     sizeof(req),
+					     &actual_len,
+					     NITROKEY_TRANSACTION_TIMEOUT,
+					     NULL,
+					     error);
+	if (!ret) {
+		g_prefix_error (error, "failed to do HOST_TO_DEVICE: ");
+		return FALSE;
+	}
+	if (actual_len != sizeof(req)) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "only wrote %" G_GSIZE_FORMAT "bytes", actual_len);
+		return FALSE;
+	}
+
+	/* get response */
+	memset (&res, 0x00, sizeof(res));
+	ret = g_usb_device_control_transfer (usb_device,
+					     G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					     G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					     G_USB_DEVICE_RECIPIENT_INTERFACE,
+					     0x01,
+					     0x0300,
+					     0x0002,
+					     (const guint8 *) &res,
+					     sizeof(res),
+					     &actual_len,
+					     NITROKEY_TRANSACTION_TIMEOUT,
+					     NULL,
+					     error);
+	if (!ret) {
+		g_prefix_error (error, "failed to do DEVICE_TO_HOST: ");
+		return FALSE;
+	}
+	if (actual_len != sizeof(res)) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "only wrote %" G_GSIZE_FORMAT "bytes", actual_len);
+		return FALSE;
+	}
+	_dump_to_console ("response", (const guint8 *) &res, sizeof(res));
+
+	/* verify this is the answer to the question we asked */
+	if (GUINT32_FROM_LE (res.last_command_crc) != crc_tmp) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "got response CRC %x, expected %x",
+			     GUINT32_FROM_LE (res.last_command_crc), crc_tmp);
+		return FALSE;
+	}
+
+	/* verify the response checksum */
+	crc_tmp = _stm_crc32 ((const guint8 *) &res, sizeof(res) - 4);
+	if (GUINT32_FROM_LE (res.crc) != crc_tmp) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "got packet CRC %x, expected %x",
+			     GUINT32_FROM_LE (res.crc), crc_tmp);
+		return FALSE;
+	}
+
+	/* copy out the payload */
+	if (buf_out != NULL)
+		memcpy (buf_out, &res.payload, buf_out_sz);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+nitrokey_execute_cmd_full (GUsbDevice *usb_device, guint8 command,
+			   const guint8 *buf_in, gsize buf_in_sz,
+			   guint8 *buf_out, gsize buf_out_sz,
+			   guint retries, guint retry_delay,
+			   GCancellable *cancellable, GError **error)
+{
```

Comment:
This is an unwieldy number of parameters to pass in a function.  Could you pack some of these into a structure and pass that structure instead?  Maybe a `NitrokeyTransaction`?